lib/std/thread.zig

```diff
@@ -252,12 +252,37 @@ pub const Thread = struct {
                 }
             }
             fn posixThreadMain(ctx: ?*c_void) callconv(.C) ?*c_void {
-                if (@sizeOf(Context) == 0) {
-                    _ = startFn({});
-                    return null;
-                } else {
-                    _ = startFn(@ptrCast(*const Context, @alignCast(@alignOf(Context), ctx)).*);
-                    return null;
+                const arg = if (@sizeOf(Context) == 0) {} else @ptrCast(*Context, @alignCast(@alignOf(Context), ctx)).*;
+
+                switch (@typeInfo(@TypeOf(startFn).ReturnType)) {
+                    .NoReturn => {
+                        startFn(arg);
+                    },
+                    .Void => {
+                        startFn(arg);
+                        return null;
+                    },
+                    .Int => |info| {
+                        if (info.bits != 8) {
+                            @compileError(bad_startfn_ret);
+                        }
+                        // pthreads don't support exit status, ignore value
+                        _ = startFn(arg);
```

Comment:
I think there's a way to make this work. I believe the `?*c_void` is not actually read by pthreads and is interpreted as an integer. So we could use `@intToPtr` here on the exit code. You can check `strace` to be sure.

Reply:
Yeah, that `?*c_void` return value is stored in `struct pthread_t` for use later with `pthread_join()` but ignored otherwise. I checked glibc and NetBSD sources, they both do that.

`strace` on a simple C test program doesn't show anything bad, so I changed this part to cast the `u8` value to `?*c_void`. Thread return values are not used in zig std yet as far as I can tell, but it could be exposed through `Thread.wait()`.

Reply:
Here, I'll show you what I mean:

```zig
const std = @import("std");

pub fn main() anyerror!void {
    const t = try std.Thread.spawn({}, start);
    t.wait();
}

fn start(context: void) u8 {
    return 123;
}
```

Here's what it looks like without pthreads:

```
[nix-shell:~/dev/zig/build]$ ./zig build-exe test3.zig 

[nix-shell:~/dev/zig/build]$ strace -f ./test3
execve("./test3", ["./test3"], 0x7ffd03c354c8 /* 128 vars */) = 0
arch_prctl(ARCH_SET_FS, 0x233b00)       = 0
rt_sigaction(SIGSEGV, {sa_handler=0x22a990, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART|SA_RESETHAND|SA_SIGINFO, sa_restorer=0x2043b0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGILL, {sa_handler=0x22a990, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART|SA_RESETHAND|SA_SIGINFO, sa_restorer=0x2043b0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGBUS, {sa_handler=0x22a990, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART|SA_RESETHAND|SA_SIGINFO, sa_restorer=0x2043b0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
mmap(NULL, 16785408, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f65bb6e4000
mprotect(0x7f65bb6e5000, 16781312, PROT_READ|PROT_WRITE) = 0
clone(child_stack=0x7f65bc6e4ff8, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID|0x400000strace: Process 22516 attached
, parent_tid=[22516], tls=0x7f65bc6e5020, child_tidptr=0x7f65bc6e5000) = 22516
[pid 22515] futex(0x7f65bc6e5000, FUTEX_WAIT, 22516, NULL <unfinished ...>
[pid 22516] exit(123)                   = ?
[pid 22515] <... futex resumed>)        = 0
[pid 22516] +++ exited with 123 +++
munmap(0x7f65bb6e4000, 16785408)        = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

Note that **the OS kernel sees exit(123)**. Now (with the patch merged in a branch) and using `-lc`:

```
[nix-shell:~/dev/zig/build]$ ./zig build-exe test3.zig -lc

[nix-shell:~/dev/zig/build]$ strace -f ./test3 
execve("./test3", ["./test3"], 0x7ffdbdb1dbd8 /* 128 vars */) = 0

(omitting a bunch of irrelevant strace)

[pid 22588] set_robust_list(0x7f979825e9e0, 24) = 0
[pid 22588] madvise(0x7f979725f000, 16752640, MADV_DONTNEED) = 0
[pid 22588] exit(0)                     = ?
[pid 22587] <... futex resumed>)        = 0
[pid 22588] +++ exited with 0 +++
munmap(0x7f979725e000, 16785408)        = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

Here you can see the thread actually calls `exit(0)` instead of `exit(123)`, so this is not accomplishing the goal.

I think the kernel probably ignores that value though. Maybe it doesn't matter. Looking at the documentation more, I see that the "retval" of `pthread_exit` is what is returned by `pthread_wait`. Zig's abstraction of that is `wait` and it does not offer a return value, which I think is reasonable. So I think my suggestion on your patch here was a red herring. Sorry about that.

Merging now!